### PR TITLE
tbb: Split into tbb_2020_3 and tbb_2021_8

### PR DESCRIPTION
**Pull request by davidak**
###### Description of changes

This is based on PR #214762.

For the new release 2021.8, see
https://www.intel.com/content/www/us/en/developer/articles/release-notes/intel-oneapi-threading-building-blocks-release-notes.html
https://github.com/oneapi-src/oneTBB/releases/tag/v2021.5.0
https://github.com/oneapi-src/oneTBB/releases/tag/v2021.6.0
https://github.com/oneapi-src/oneTBB/releases/tag/v2021.7.0

> Due to the significant breakage due to the update to TBB 2021.8, instead split the tbb package into tbb_2020_3 and tbb_2021_8, with the default tbb aliased to tbb_2020_3 in order to minimize breakage.
> 
> The reason I went for this change is that I'd like to package [CCTag](https://github.com/alicevision/CCTag) which depends on TBB 2021. To me this solution seems like a good way to bridge the time until the packages incompatible with TBB 2021.8, but I'm not sure if there are any guidelines regarding splitting packages like this. So please comment if this discouraged for some reason.
> 
> I hope I got the split right. In particular, I'm wondering if there's a better solution for sharing `meta`. Also, I'm open to suggestions regarding the versioned package naming scheme.

Continuing https://github.com/NixOS/nixpkgs/pull/215689

cc co-author @hesiod

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] i686-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by mweinelt** [comment]
Stop spamming me ASAP. I can't deal with it. Remove my account name from [`bede796` (#216991)](https://github.com/NixOS/nixpkgs/pull/216991/commits/bede79603b4acdb8833f5892d2d2b8d26c2f5e3b) NOW.

Screenshot 2023-02-18 at 19-32-10 Build software better together

**Comment by davidak** [comment]
@mweinelt done!

I was confused what you meant as i did not create the commit.

Sorry! :disappointed:

**Comment by davidak** [comment]
Result of `nixpkgs-review pr 216991` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>tbb_2020_3</li>
    <li>tbb_2021_8</li>
  </ul>
</details>

**Comment by davidak** [comment]
Result of `nixpkgs-review pr 216991` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>tbb_2020_3</li>
    <li>tbb_2021_8</li>
  </ul>
</details>

**Comment by davidak** [comment]
Result of `nixpkgs-review pr 216991` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>tbb_2020_3</li>
    <li>tbb_2021_8</li>
  </ul>
</details>

**Comment by davidak** [comment]
Result of `nixpkgs-review pr 216991` run on i686-linux [1](https://github.com/Mic92/nixpkgs-review)

**Comment by davidak** [comment]
Result of `nixpkgs-review pr 216991` run on i686-linux [1](https://github.com/Mic92/nixpkgs-review)

**Comment by davidak** [comment]
i'm building both packages for i686-linux, but it will take some time. it has to build staging from scratch including 3 gcc versions!

it would be great if someone can test on aarch64-linux. it fails with binfmt emulation

**Comment by davidak** [comment]
tbb_2020_3 build works on i686-linux (32-bit), but tbb_2021_8 fails (https://github.com/oneapi-src/oneTBB/issues/1030) as reported before (https://github.com/NixOS/nixpkgs/pull/214762#issuecomment-1418288171)

**Comment by davidak** [comment]
Result of `nixpkgs-review pr 216991` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

**Comment by davidak** [comment]
My system now builds 4768 packages on x86_64-linux...

**Comment by trofi** [comment]
> @trofi i removed your gcc13 fix as they don't support gcc13 (https://github.com/NixOS/nixpkgs/commit/0e1c9977e9325f4dec21c46a80badf3b525aed48). when they do, it should not be needed

That is a bit unusual reason to remove a patch that fixes a) guaranteed build failure on a compiler that will soon release b) accepted upstream. Does patch fail to apply or cause other issues?

File: pkgs/development/libraries/tbb/2020_3.nix [hunk 1]
```diff
@@ -0,0 +1,85 @@
+{ lib
+, gcc9Stdenv
+, fetchurl
+, fetchFromGitHub
+, fixDarwinDylibNames
+}:
+
+let
+  stdenv = gcc9Stdenv;
```
Comment (trofi) [inline pkgs/development/libraries/tbb/2020_3.nix:9]: `boost` is a c++ library that encodes compiler version into it's function names to make sure ABI does not rebak yb accident. I think it is crucial to use default compiler here (and frankly in every frequently used library C++ library in `nixpkgs`).

I would expect numerous build breakages when attempting to use it with a compiler different the one boost was built. Things will probably break even more if you build it on clang-based system.

Why do you need `gcc-9` here?

File: pkgs/development/libraries/tbb/2021_8.nix [hunk 2]
```diff
@@ -0,0 +1,41 @@
+{ lib
+, gcc11Stdenv
+, fetchFromGitHub
+, fetchpatch
+, cmake
+}:
+
+let
+  stdenv = gcc11Stdenv;
```
Comment (trofi) [inline pkgs/development/libraries/tbb/2021_8.nix:9]: Same as above, why not use default compiler here? Using non-default compiler will probably break many more things.

**Review by trofi — CHANGES_REQUESTED** [review]

File: pkgs/development/libraries/tbb/2020_3.nix (hunk 1 above)
Comment (davidak) [reply pkgs/development/libraries/tbb/2020_3.nix:9]: The build fails with default gcc. See https://github.com/NixOS/nixpkgs/pull/215689#issuecomment-1435060127

GCC 9 is the latest supported version, so i thought we should use that.
Not sure how to handle clang here.

I don't know how this is usually handled in nixpkgs. It's the first time i have to deal with this.

File: pkgs/development/libraries/tbb/2021_8.nix (hunk 2 above)
Comment (davidak) [reply pkgs/development/libraries/tbb/2021_8.nix:9]: Same reason as above. It is not supported. And the build failed.

File: pkgs/development/libraries/tbb/2020_3.nix (hunk 1 above)
Comment (chuangzhu) [reply pkgs/development/libraries/tbb/2020_3.nix:9]: We had a patch that fixed the build for 2020.3. It works fine and is accepted by the upstream. Why do you want to change that?

Comment (trofi) [reply pkgs/development/libraries/tbb/2020_3.nix:9]: > The default compiler version is not supported.

The way I understand "not supported" is when major work is required  to adapt the software to the environment. It's a very rare occasion when old `gcc` just works and new `gcc` requires major work to adapt. `gcc` usually provides flags to get behaviour of an old compiler even on new `gcc` versions (be it disabling warnings, forcing an older version of C standard and whatnot). Mixing compiler versions and getting unique bugs is a real possibilty to get undebuggable problems.

Looking at https://github.com/oneapi-src/oneTBB/pull/866 the fix is trivial to backport. I would say it does not warrant using an outdated compiler. If that is the only problem I can help tweaking the build to succeed on modern `gcc`.

> Not sure how to handle clang here.

Usually it requires a more complex condition, like `if stdenv.isClang then stdenv else gcc9Stdenv`. But once you in that territory it's frequently easier to just fix the package.

Comment (davidak) [reply pkgs/development/libraries/tbb/2020_3.nix:9]: @chuangzhu said "GCC 12 isn't supported in v2021.8.0" so the obvious solution for me was to use the supported compiler, for both packages.

If it's consensus we can ignore upstream support status and better use the default version, we can try to get it working. I appreciate your help! I'm usually not dealing with such low level problems.

@trofi thanks for the explanation!

**Comment by davidak** [comment]
@trofi i applied the patch from https://github.com/oneapi-src/oneTBB/pull/866 but the build of tbb_2021_8 still fails.

```
tbb> [ 65%] Building CXX object test/CMakeFiles/conformance_tick_count.dir/conformance/conformance_tick_count.cpp.o
tbb> In file included from /nix/store/0spl4agq94fairydh94xjmsm47zfk684-gcc-12.2.0/include/c++/12.2.0/atomic:41,
tbb>                  from /build/source/test/common/doctest.h:2990,
tbb>                  from /build/source/test/common/test.h:32,
tbb>                  from /build/source/test/tbb/test_concurrent_monitor.cpp:26:
tbb> In member function 'void std::__atomic_base<_IntTp>::store(__int_type, std::memory_order) [with _ITp = bool]',
tbb>     inlined from 'void std::atomic<bool>::store(bool, std::memory_order)' at /nix/store/0spl4agq94fairydh94xjmsm47zfk684-gcc-12.2.0/include/c++/12.2.0/atomic:104:20,
tbb>     inlined from 'void tbb::detail::r1::concurrent_monitor_base<Context>::abort_all_relaxed() [with Context = long unsigned int]' at /build/source/test/tbb/../../src/tbb/concurrent_monitor.h:430:53:
tbb> /nix/store/0spl4agq94fairydh94xjmsm47zfk684-gcc-12.2.0/include/c++/12.2.0/bits/atomic_base.h:464:25: error: 'void __atomic_store_1(volatile void*, unsigned char, int)' writing 1 byte into a region of size 0 overflows the destination [-Werror=stringop-overflow=]
tbb>   464 |         __atomic_store_n(&_M_i, __i, int(__m));
tbb>       |         ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
tbb> compilation terminated due to -Wfatal-errors.
tbb> cc1plus: all warnings being treated as errors
tbb> make[2]: *** [test/CMakeFiles/test_concurrent_monitor.dir/build.make:76: test/CMakeFiles/test_concurrent_monitor.dir/tbb/test_concurrent_monitor.cpp.o] Error 1
tbb> make[1]: *** [CMakeFiles/Makefile2:2234: test/CMakeFiles/test_concurrent_monitor.dir/all] Error 2
```

**Comment by davidak** [comment]
A rebase added again many unrelated commits and added notification for unrelated people. Sorry

**Comment by trofi** [comment]
I removed extra reviewers.

**Pull request closed without merging**